tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/AUC.java

```diff
@@ -0,0 +1,1074 @@
+      Operand<? extends TNumber> labels,
+      Operand<? extends TNumber> predictions,
+      Operand<? extends TNumber> sampleWeights) {
+    Ops tf = getTF();
+    Operand<T> tLabels = cast(tf, labels, type);
+    Operand<T> tPredictions = cast(tf, predictions, type);
+    Operand<T> tSampleWeights = sampleWeights != null ? cast(tf, sampleWeights, type) : null;
+    List<Op> updateOperations = new ArrayList<>();
+    Map<ConfusionMatrixEnum, Assign<T>> varInitializers = Collections.EMPTY_MAP;
+    if (!initialized) {
+      varInitializers = build(tPredictions.shape());
+    }
+    if (isMultiLabel() || getLabelWeights() != null) {
+      // labels should have shape (number of examples, number of labels).
+      List<SymbolicShape<? extends TNumber>> symbols = new ArrayList<>();
+      symbols.add(new SymbolicShape<>(tLabels, "N", "L"));
+      if (isMultiLabel()) {
+        // TP, TN, FP, and FN should all have shape
+        // (number of thresholds, number of labels).
+        symbols.add(new SymbolicShape<>(truePositives, "T", "L"));
+        symbols.add(new SymbolicShape<>(falsePositives, "T", "L"));
+        symbols.add(new SymbolicShape<>(trueNegatives, "T", "L"));
+        symbols.add(new SymbolicShape<>(falseNegatives, "T", "L"));
+      }
+      if (getLabelWeights() != null) {
+        symbols.add(new SymbolicShape<>(getLabelWeights(), "L"));
+      }
+      updateOperations.addAll(
+          MetricsHelper.assertShapes(tf, symbols, "Number of labels is not consistent."));
+    }
+
+    Map<ConfusionMatrixEnum, Variable<T>> confusionMatrix = new HashMap<>();
+    confusionMatrix.put(ConfusionMatrixEnum.TRUE_POSITIVES, truePositives);
+    confusionMatrix.put(ConfusionMatrixEnum.FALSE_POSITIVES, falsePositives);
+    confusionMatrix.put(ConfusionMatrixEnum.TRUE_NEGATIVES, trueNegatives);
+    confusionMatrix.put(ConfusionMatrixEnum.FALSE_NEGATIVES, falseNegatives);
+
+    // Only forward labelWeights to update_confusion_matrix_variables when
+    // multiLabel is false. Otherwise the averaging of individual label AUCs is
+    // handled in AUC.result
+    updateOperations.addAll(
+        MetricsHelper.updateConfusionMatrixVariables(
+            tf,
+            confusionMatrix,
+            varInitializers,
+            tLabels,
+            tPredictions,
+            tf.constant(thresholds),
+            null,
+            null,
+            tSampleWeights,
+            isMultiLabel(),
+            isMultiLabel() ? null : getLabelWeights()));
+    return updateOperations;
+  }
+
+  /**
+   * Gets the input with all positive numbers. Negative numbers are set to 0.
+   *
+   * @param input the input
+   * @return the input with all positive numbers.
+   */
+  private Operand<T> positive(Operand<T> input) {
+    return getTF().math.maximum(input, cast(getTF(), getTF().constant(0), input.type()));
+  }
+
+  /**
+   * Gets an operand that determines whether the input consists of each value is greater than 0.
```

Comment:
Actually, it's "the truth value of whether `input > 0`, element-wise. (Here and in the `@return`.)

Reply:
OK